File: python/aoc/domain/compute/intcode_computer.py
from domain.compute.instructions.compare import Compare
from domain.compute.instructions.display import Display
from domain.compute.instructions.halt import Halt
from domain.compute.instructions.jump import Jump
from domain.compute.instructions.math import Math
from domain.compute.instructions.parameter_mode import ParameterMode
from domain.compute.instructions.put import Put
from domain.compute.invalid_intcode_program import InvalidIntcodeProgram
from domain.compute.opcodes import Opcodes


class IntcodeComputer(object):
    def __init__(self):
        self._memory = None
        self._input = 0
        self._instructionPointer = None
        self._Output = None
    @property
    def Output(self):
        return self._Output
    @Output.setter
    def Output(self, value):
        self._Output = value

    def RunProgram(self, program, userinput):
        self.Initialize(program, userinput)
        self._instructionPointer = 0

        #Main computer logic
        while(self._instructionPointer < len(program)):
            instruction = self.GetNextInstruction()
            if isinstance(instruction, Halt):
                return self._memory
            self._instructionPointer = self.ExecuteInstruction(instruction)

        raise InvalidIntcodeProgram("No halt instruction at end of program")

    def GetNextInstruction(self):
        opcode = self._memory[self._instructionPointer]
        opcodeEnum = Opcodes(opcode % 100)
        match opcodeEnum:
            case Opcodes.Halt:
                return Halt()
            case Opcodes.Display:
                return Display()
            case Opcodes.Put:
                return Put(opcode)
            case Opcodes.JumpIfTrue:
                return Jump(opcode, lambda: True)
            case Opcodes.JumpIfFalse:
                return Jump(opcode, lambda: False)
            case Opcodes.LessThan:
                return Jump(opcode, lambda x, y: x < y)
            case Opcodes.Equals:
                return Jump(opcode, lambda x, y: x == y)
            case Opcodes.Add:
                return Math(opcode, lambda x, y: x + y)
            case Opcodes.Multiply:
                return Math(opcode, lambda x, y: x * y)

    def ExecuteInstruction(self, instruction):
        if isinstance(instruction, Math):
            self.ExecuteMathInstruction(instruction)
        if isinstance(instruction, Display):
            self.ExecuteDisplayInstruction(instruction)
        if isinstance(instruction, Put):
            self.ExecutePutInstruction(instruction)
        if isinstance(instruction, Compare):
            self.ExecuteCompareInstruction(instruction)
        if isinstance(instruction, Jump):
            return self.ExecuteJumpInstruction(instruction)
        return self._instructionPointer + instruction.Length

    def ExecuteJumpInstruction(self, instruction):
        parameter1 = self.GetParameterValue(instruction, 1)
        parameter2 = self.GetParameterValue(instruction, 2)
        if instruction.ShouldJump(parameter1):
            return parameter2
        return (self._instructionPointer + instruction.Length)

    def ExecuteCompareInstruction(self, instruction):
        parameter1 = self.GetParameterValue(instruction, 1)
        parameter2 = self.GetParameterValue(instruction, 2)
        destinationAddress = self._memory[(self._instructionPointer + 3)]
        self._memory[destinationAddress] = (1 if instruction.CompareFunction(parameter1, parameter2) else 0)

    def ExecuteMathInstruction(self, instruction):
        parameter1 = self.GetParameterValue(instruction, 1)
        parameter2 = self.GetParameterValue(instruction, 2)
        instructionValue = instruction.MathOperation(parameter1, parameter2)
        destinationAddress = self._memory[(self._instructionPointer + 3)]
        self._memory[destinationAddress] = instructionValue

    def ExecuteDisplayInstruction(self, instruction):
        self._Output.append(self._memory[self._memory[(self._instructionPointer + 1)]])

    def ExecutePutInstruction(self, instruction):
        if (self._input == None):
            raise InvalidIntcodeProgram("This program expects input from user and none was given")
        if (instruction.ParameterModes[0] == ParameterMode.Immediate):
            self._memory[(self._instructionPointer + 1)] = self._input
        else:
            self._memory[self._memory[(self._instructionPointer + 1)]] = self._input

    def GetParameterValue(self, instruction, parameterPosition):
        if (instruction.ParameterModes[(parameterPosition - 1)] == ParameterMode.Immediate):
            return self._memory[(self._instructionPointer + parameterPosition)]
        return self._memory[self._memory[(self._instructionPointer + parameterPosition)]]

    def Initialize(self, program, userInput):
        self._input = userInput
        #Don't surprise the user and change the incoming program
        self._Output = []
        self._memory = program.copy()

File: dotnet/Aoc/Aoc.Spaceship/Compute/IntcodeComputer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Aoc.Spaceship.Compute.Instructions;
using Math = Aoc.Spaceship.Compute.Instructions.Math;

namespace Aoc.Spaceship.Compute
{
    public class IntcodeComputer {
        private int[]? _memory;
        private List<int> _input;
        private int _instructionPointer;
        public List<int>? Output { get; set;  }

        public int[] RunProgram(int[] program, int? input = null)
        {
            return RunProgram(program, new[] {input.Value});
            Initialize(program, input);

            //Main computer logic
            _instructionPointer = 0;
            while (_instructionPointer < _memory.Length)
            {
                var instruction = GetNextInstruction();
                if (instruction is Halt)
                    return _memory;
                _instructionPointer = ExecuteInstruction(instruction);
            }

            //We should always see a halt operation at the end of the program
            throw new InvalidIntcodeProgram("No halt instruction at end of program");
        }

        public int[] RunProgram(int[] program, int[] input)
        {
            Initialize(program, input.ToList());

            //Main computer logic
            _instructionPointer = 0;
            while (_instructionPointer < _memory.Length)
            {
                var instruction = GetNextInstruction();
                if (instruction is Halt)
                    return _memory;
                _instructionPointer = ExecuteInstruction(instruction);
            }

            //We should always see a halt operation at the end of the program
            throw new InvalidIntcodeProgram("No halt instruction at end of program");
        }
l  h     private Instruction GetNextInstruction()
        {
            var rawOpcode = _memory[_instructionPointer];
            Opcodes opcode = (Opcodes) (rawOpcode % 100);
            Instruction instruction;
            switch (opcode)
            {
                case Opcodes.Halt:
                    instruction = new Halt();
                    break;
                case Opcodes.Display:
                    instruction = new Display();
                    break;
                case Opcodes.Put:
                    instruction = new Put(rawOpcode);
                    break;
                case Opcodes.JumpIfTrue:
                    instruction = new Jump(rawOpcode,() => true);
                    break;
                case Opcodes.JumpIfFalse:
                    instruction = new Jump(rawOpcode,() => false);
                    break;
                case Opcodes.LessThan:
                    instruction = new Compare(rawOpcode, compareFunction: (x, y) => x < y);
                    break;
                case Opcodes.Equals:
                    instruction = new Compare(rawOpcode, compareFunction: (x, y) => x == y);
                    break;
                case Opcodes.Add:
                    instruction = new Math(rawOpcode, mathOperation: (x, y) => x + y);
                    break;
                case Opcodes.Multiply:
                    instruction = new Math(rawOpcode, mathOperation: (x, y) => x * y);
                    break;
                default:
                    throw new InvalidIntcodeProgram($"Opcode {opcode} unknown");
            }
            if (_memory.Length < _instructionPointer + instruction.Length)
                throw new InvalidIntcodeProgram("Last instruction is incomplete");
            return instruction;
        }

        private int ExecuteInstruction(Instruction instruction)
        {
            switch (instruction)
            {
                case Math mathInstruction:
                    ExecuteInstruction(mathInstruction);
                    break;
                case Display displayInstruction:
                    ExecuteInstruction(displayInstruction);
                    break;
                case Put putInstruction:
                    ExecuteInstruction(putInstruction);
                    break;
                case Jump jumpInstruction:
                    return ExecuteInstruction(jumpInstruction);
                case Compare compareInstruction:
                    ExecuteInstruction(compareInstruction);
                    break;
                default: 
                    throw new InvalidIntcodeProgram($"Unknown instruction {instruction}");
            }
            return _instructionPointer + instruction.Length;
        }

        private int ExecuteInstruction(Jump instruction)
        {
            int parameter1 = GetParameterValue(instruction, 1);
            int parameter2 = GetParameterValue(instruction, 2);
            if (instruction.ShouldJump(parameter1))
                return parameter2;
            return _instructionPointer + instruction.Length;
        }

        private void ExecuteInstruction(Compare instruction)
        {
            int parameter1 = GetParameterValue(instruction, 1);
            int parameter2 = GetParameterValue(instruction, 2);
            var destinationAddress = GetParameterValue(instruction, 3);
            _memory[destinationAddress] = instruction.CompareFunction(parameter1, parameter2) ? 1 : 0;
        }

        private void ExecuteInstruction(Math instruction)
        {
            int parameter1 = GetParameterValue(instruction, 1);
            int parameter2 = GetParameterValue(instruction, 2);
            var instructionValue = instruction.MathOperation(parameter1, parameter2);
            var destinationAddress = GetParameterValue(instruction, 3);
            _memory[destinationAddress] = instructionValue;
        }

        private void ExecuteInstruction(Display instruction)
        {
            Output.Add(GetParameterValue(instruction, 1));
        }

        private void ExecuteInstruction(Put instruction)
        {
            if (!_input.HasValue)
                throw new InvalidIntcodeProgram("This program expects input from user and none was given");
            if (instruction.ParameterModes[0] == ParameterMode.Immediate)
                _memory[_instructionPointer + 1] = _input.Value;
            else
                _memory[_memory[_instructionPointer + 1]] = _input.Value;
        }

        private int GetParameterValue(Instruction instruction, int parameterPosition)
        {
            if (instruction.ParameterModes[parameterPosition - 1] == ParameterMode.Immediate)
                return _memory[_instructionPointer + parameterPosition];
            return _memory[_memory[_instructionPointer + parameterPosition]];
        }

        private void Initialize(int[] program, int[] input)
        {
            //Make sure we don't use input from previous runs
            _input = input;
            //Make sure we don't have output from previous program runs
            Output = new List<int>();
            //Don't surprise the user and make changes to the incoming program
            _memory = Array.Empty<int>();
            Array.Resize(ref _memory, program.Length);
            program.CopyTo(_memory, 0);
        }
    }
}
